docs/developers/code-style.md

```diff
@@ -220,6 +365,162 @@ where:
 - `when` - stimulus: some action against the system under test.
 - `then` - expectations/assertions.
 
+#### Tests granularity
+
+Unit tests should be as granular as possible. Try to split unit tests into smaller ones until this is impossible to do.
+
+```
+// bad
+@Test
+public void testFooBar() {
+    // when
+    final String foo = service.getFoo();
+    final String bar = service.getBar();
+    
+    // then
+    assertThat(foo).isEqualTo("foo");
+    assertThat(bar).isEqualTo("bar");
+}
+
+// good
+@Test
+public void testFoo() {
+    // when
+    final String foo = service.getFoo();
+    
+    // then
+    assertThat(foo).isEqualTo("foo");
+}
+
+@Test
+public void testBar() {
+    // when
+    final String bar = service.getBar();
+    
+    // then
+    assertThat(bar).isEqualTo("bar");
+}
+```
+
+This also applies to cases where same method is tested with different arguments inside single unit test.
+Note: This represents the replacement we have selected for parameterized testing.
+
+```
+// bad
+@Test
+public void testFooFirstSecond() {
+    // when
+    final String foo1 = service.getFoo(1);
+    final String foo2 = service.getFoo(2);
+    
+    // then
+    assertThat(foo1).isEqualTo("foo1");
+    assertThat(foo2).isEqualTo("foo2");
+}
+
+// good
+@Test
+public void testFooFirst() {
+    // when
+    final String foo1 = service.getFoo(1);
+    
+    // then
+    assertThat(foo1).isEqualTo("foo1");
+}
+
+@Test
+public void testFooSecond() {
+    // when
+    final String foo2 = service.getFoo(2);
+    
+    // then
+    assertThat(foo2).isEqualTo("foo2");
+}
+```
+
+#### Unit tests naming
+
+Name unit tests meaningfully. Test names should give brief description of what unit test tries to check.
+It is also recommended to prefix test methods with the name of method that is being tested and word `should`.
+
+```
+// bad
+@Test
+public void doSomethingTest() {
+    // when and then
+    assertThat(service.processData()).isEqualTo("result");
+}
+
+// good
+@Test
+public void processDataShouldReturnResult() {
+    // when and then
+    assertThat(service.processData()).isEqualTo("result");
+}
+```
```

Comment:
matter of taste, but as an option I suggest using naming pattern should...when... (I believe it's more common for BDD style of tests, for example, mockito use such example in their javadoc, so I assume it's a standard at some point)

also I suggest using the `systemUnderTest` or teven the shorter version `sut` in order to outline the service we cover in order not to mess up with its dependancies